airbyte-cdk/python/airbyte_cdk/sources/file_based/file_based_stream_reader.py

```diff
@@ -54,11 +54,22 @@ def filter_files_by_globs(files: List[RemoteFile], globs: List[str]) -> List[Rem
         """
         Utility method for filtering files based on globs.
         """
-        return [file for file in files if any(fnmatch(file.uri, glob) for glob in globs)]
+        if not globs:
+            return files
+
+        matches = {}
+
+        for file in files:
+            for g in globs:
+                if globmatch(file.uri, g, flags=GLOBSTAR):
+                    matches[file.uri] = file
+
+        return list(matches.values())
```

Comment:
For a function with nested that could operate on a large set and potentially highly recursive, it might be worthwhile to return an `Iterable[RemoteFile]` so we can continue doing work as more files are found. The main drawbacks would be if we require logic that requires looking at the whole set of files or if it adds a lot of complexity to what is invoking this function (which hasn't been implemented yet afaict). Curious to get your take on this?

Reply:
That makes sense to me and I've updated this to be a generator.

Some additional context - the originator of the call to this function will be from the Stream class's `list_files` method, which returns a List so that we can cache the results rather than hitting the source multiple times, if we needed to `list_files` more than once. (Since `list_files` returns a list and that's the only entrypoint into this `filter_files_by_globs`, there aren't perf improvements from making it a generator so I left it as a list, which felt more natural for deduping files.) I went back and forth on whether that caching should be on the stream's method versus the stream reader, which is what's actually making the calls to the source, but ultimately put it on the stream so that it's still easy to bypass it - e.g. in cases where the stream reader is paging through results - we don't necessarily want to block while paging. 

FWIW the existing [source_files_abstract code](https://github.com/airbytehq/airbyte/blob/d9a5e2d873a036a464c955aec7f5809c3760f432/airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/stream.py#L157) also works on the whole list (for caching) and AFAIK there haven't been OOM issues from it, but it is in the back of my mind as something that we should keep an eye on. I'm tempted to get rid of the caching as a protective measure, under the assumption that `list_files` calls should be rare so it should be unlikely that we ever do end up calling it more than once. WDYT?

Reply:
ah got it. Thanks for the context on the caching and how we're turning it into a list further upstream. Knowing all this and that the existing connector wasn't running into any memory issues, I don't have super strong opinions that one is a must over the other. 

And yeah framed the way you mention it, it doesn't feel like we need a cache here and even if its lightweight and just in the Python application code. So I'm on board with not having one

Reply:
Actually there's another consideration for listing files, which is that some sources (S3) don't allow us to sort by last-modified, so if we want to support incremental we'll have to sort ourselves, meaning we'll have to retrieve all the files first. So I think I'll leave the cache in for now.